Register `~const` preds for `Deref` adjustments in HIR typeck

compiler-errors opened the pull request:
This doesn't *do* anything yet, since `Deref` and `DerefMut` aren't constified, and we explicitly don't error on calling non-const trait methods in HIR yet -- presumably that will wait until std is re-constified. But I'm confident this logic is correct, and this (afaict?) is the only major hole left in enforcing `~const` in HIR typeck.

r? fee1-dead

fee1-dead commented:
Haven't looked at the code closely, but this could probably use a minicore test if possible.

compiler-errors committed:
Fix minicore, add tests based off of it

compiler-errors commented:
I fixed up the existing minicore, moved it to an auxiliary dependency, so we can start adding real tests based off of that minicore that exercise its behavior.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11581245037/job/32241557826) [(plain)](https://github.com/rust-lang/rust/commit/14a6865cb2dc2d5df6d422768983b6cdb97a5c58/checks/32241557826/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.116 Building wheels for collected packages: reuse
#13 3.118   Building wheel for reuse (pyproject.toml): started
#13 3.365   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.366   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.366   Stored in directory: /tmp/pip-ephem-wheel-cache-4336c3w1/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.369 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.763 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.763 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.306 Collecting virtualenv
#13 4.306 Collecting virtualenv
#13 4.377   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.473      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 33.3 MB/s eta 0:00:00
#13 4.511 Collecting distlib<1,>=0.3.7
#13 4.526   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.535      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 75.5 MB/s eta 0:00:00
#13 4.571 Collecting filelock<4,>=3.12.2
#13 4.585   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.617 Collecting platformdirs<5,>=3.9.1
#13 4.631   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.713 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.886 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 5.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      174016 kB
DirectMap2M:     5068800 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5641 files
tidy check
tidy error: Stray file with UI testing output: "/checkout/tests/ui/traits/const-traits/effects/minicore.stderr"
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Oct 29 19:49:31 UTC 2024
  network time: Tue, 29 Oct 2024 19:49:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #132301) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fee1-dead commented on compiler/rustc_hir_typeck/src/place_op.rs:
```diff
@@ -289,13 +289,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                            );
-                        };
-                        *deref = OverloadedDeref { mutbl, span: deref.span };
+                        if let ty::Ref(_, _, mutbl) = *method.sig.output().kind() {
+                            *deref = OverloadedDeref { mutbl, span: deref.span };
+                        }
```
Why was this span_bug removed?

compiler-errors replied:
It's kinda pointless. I think I mixed up a bit of the refactoring from #132301, but all it's doing is asserting something we've explicitly decided not to care about (which is detecting invalid libcore items, https://github.com/rust-lang/compiler-team/issues/620).

compiler-errors replied:
But whatever, I'll revert it.

fee1-dead commented on compiler/rustc_middle/src/ty/adjustment.rs:
```diff
@@ -124,18 +125,16 @@ pub struct OverloadedDeref {
 
 impl OverloadedDeref {
     /// Get the zst function item type for this method call.
```
This doc should be updated

compiler-errors committed:
Register const preds for Deref adjustments in HIR typeck

compiler-errors committed:
Make sure to enforce ~const DerefMut on mutability fixup

compiler-errors commented:
@rustbot ready

fee1-dead commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 8ed8f22d8d1f844792a4a70b787d182f299202b7 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 8ed8f22d8d1f844792a4a70b787d182f299202b7 -->
<!-- homu: {"type":"Approved","sha":"8ed8f22d8d1f844792a4a70b787d182f299202b7","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
failed in https://github.com/rust-lang/rust/pull/132551#issuecomment-2453336111

```

warning: function `const_eval_select` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:465:10
   |
   |
LL | const fn const_eval_select<ARG: Tuple, F, G, RET>(

error: linking with `cc` failed: exit status: 1
   |
   |
   = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/bin:/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin:/Users/runner/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/runner/.cargo/bin:/usr/local/opt/curl/bin:/usr/local/bin:/usr/local/sbin:/Users/runner/bin:/Users/runner/.yarn/bin:/Users/runner/Library/Android/sdk/tools:/Users/runner/Library/Android/sdk/platform-tools:/Library/Frameworks/Python.framework/Versions/Current/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/bin:/bin:/usr/sbin:/sbin:/Users/runner/.dotnet/tools" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-Wl,-exported_symbols_list" "-Wl,/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustclDu9Z6/list" "/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustclDu9Z6/symbols.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.minicore.356631b40ca0b657-cgu.0.rcgu.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.0lca3mr32tz9pvmmfm5apoe3h.rcgu.rmeta" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary" "-o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/libminicore.dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-install_name" "-Wl,@rpath/libminicore.dylib" "-nodefaultlibs"
   = note: Undefined symbols for architecture arm64:
             "dyld_stub_binder", referenced from:
                 <initial-undefines>
           ld: symbol(s) not found for architecture arm64
           clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

@bors r-

Uh can we get a ~~doctor~~ orchard farmer in the house?

workingjubilee commented:
@bors try

bors commented:
:hourglass: Trying commit 8ed8f22d8d1f844792a4a70b787d182f299202b7 with merge b8cc2dc76fbee90868693d9e0739dd4d30983ad6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8ed8f22d8d1f844792a4a70b787d182f299202b7","merge_sha":"b8cc2dc76fbee90868693d9e0739dd4d30983ad6"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11649791276/job/32437734094) [(plain)](https://github.com/rust-lang-ci/rust/commit/b8cc2dc76fbee90868693d9e0739dd4d30983ad6/checks/32437734094/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (49251/50256)
Updating files:  99% (49754/50256)
Updating files: 100% (50256/50256)
Updating files: 100% (50256/50256), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/opt/homebrew/bin/git log -1 --format=%H
b8cc2dc76fbee90868693d9e0739dd4d30983ad6
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [ui] tests/ui/traits/const-traits/effects/minicore-works.rs stdout ----

error: auxiliary build of "/Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs" failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" "/Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/work/rust/rust/vendor" "--sysroot" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2" "--target=aarch64-apple-darwin" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-Znext-solver" "--crate-type" "dylib" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary"
--- stderr -------------------------------
warning: unused variable: `rhs`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:39:18
   |
---

warning: unused variable: `x`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:45:9
   |
LL |     let x = 42_i32 + 43_i32;
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: unused variable: `x`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:49:9
   |
   |
LL |     let x = 42_i32 + 43_i32;
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: unused variable: `bytes`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:140:28
   |
   |
LL |     const fn from_ne_bytes(bytes: [u8; size_of::<Self>()]) -> Self {
   |                            ^^^^^ help: if this is intentional, prefix it with an underscore: `_bytes`
warning: unused variable: `other`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:316:18
   |
   |
LL |     fn eq(&self, other: &str) -> bool {
   |                  ^^^^^ help: if this is intentional, prefix it with an underscore: `_other`
warning: unused variable: `value`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:415:18
   |
   |
LL |     const fn new(value: T) -> RefCell<T> {
   |                  ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
warning: unused variable: `arg`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:466:5
   |
LL |     arg: ARG,
---

warning: unused variable: `called_at_rt`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:468:5
   |
LL |     called_at_rt: G,
   |     ^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_called_at_rt`

warning: type `ControlFlow<<Self as Try>::Residual, <Self as Try>::Output>` is more private than the item `Try::branch`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:62:5
   |
LL |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method `Try::branch` is reachable at visibility `pub`
   |
note: but type `ControlFlow<<Self as Try>::Residual, <Self as Try>::Output>` is only usable at visibility `pub(crate)`
  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:71:1
LL | enum ControlFlow<B, C = ()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `#[warn(private_interfaces)]` on by default

---
   |
LL | const fn bar() {
   |          ^^^

warning: variants `Continue` and `Break` are never constructed
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:73:5
LL | enum ControlFlow<B, C = ()> {
   |      ----------- variants in this enum
   |      ----------- variants in this enum
LL |     #[lang = "Continue"]
LL |     Continue(C),
   |     ^^^^^^^^
LL |     #[lang = "Break"]
LL |     Break(B),

warning: function `size_of` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:130:10
   |
   |
LL | const fn size_of<T>() -> usize {


warning: associated functions `repeat_u8` and `from_ne_bytes` are never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:136:14
LL | impl usize {
   | ---------- associated functions in this implementation
LL |     #[rustc_allow_incoherent_impl]
LL |     #[rustc_allow_incoherent_impl]
LL |     const fn repeat_u8(x: u8) -> usize {
...
...
LL |     const fn from_ne_bytes(bytes: [u8; size_of::<Self>()]) -> Self {

warning: function `panic_display` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:146:10
   |
---
   |      ------ variants in this enum
LL |     #[lang = "None"]
LL |     None,
   |     ^^^^
LL |     #[lang = "Some"]
LL |     Some(T),

warning: methods `as_ref` and `as_mut` are never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:231:14
   |
   |
LL | impl<T> Option<T> {
   | ----------------- methods in this implementation
LL |     const fn as_ref(&self) -> Option<&T> {
...
LL |     const fn as_mut(&mut self) -> Option<&mut T> {
   |              ^^^^^^


warning: function `as_deref` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:248:10
   |
LL | const fn as_deref<T>(opt: &Option<T>) -> Option<&T::Target>

warning: enum `Result` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:283:6
   |
---
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:371:14
   |
LL | impl<T> Option<T> {
   | ----------------- methods in this implementation
LL |     const fn as_pin_ref(self: Pin<&Self>) -> Option<Pin<&T>> {
...
...
LL |     const fn as_pin_mut(self: Pin<&mut Self>) -> Option<Pin<&mut T>> {

warning: struct `RefCell` is never constructed
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:410:8
   |
---
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:415:14
   |
LL | impl<T> RefCell<T> {
   | ------------------ associated function in this implementation
LL |     const fn new(value: T) -> RefCell<T> {

warning: method `borrow` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:420:8
   |
---
LL | struct Ref<'b, T: ?Sized + 'b> {
   |        --- fields in this struct
LL |     value: *const T,
   |     ^^^^^
LL |     borrow: &'b UnsafeCell<()>,

warning: function `const_eval_select` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:465:10
   |
   |
LL | const fn const_eval_select<ARG: Tuple, F, G, RET>(

error: linking with `cc` failed: exit status: 1
   |
   |
   = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/bin:/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin:/Users/runner/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/runner/.cargo/bin:/usr/local/opt/curl/bin:/usr/local/bin:/usr/local/sbin:/Users/runner/bin:/Users/runner/.yarn/bin:/Users/runner/Library/Android/sdk/tools:/Users/runner/Library/Android/sdk/platform-tools:/Library/Frameworks/Python.framework/Versions/Current/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/bin:/bin:/usr/sbin:/sbin:/Users/runner/.dotnet/tools" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-Wl,-exported_symbols_list" "-Wl,/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustcheQ4VZ/list" "/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustcheQ4VZ/symbols.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.minicore.356631b40ca0b657-cgu.0.rcgu.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/minicore.0lca3mr32tz9pvmmfm5apoe3h.rcgu.rmeta" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary" "-o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-works/auxiliary/libminicore.dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-install_name" "-Wl,@rpath/libminicore.dylib" "-nodefaultlibs"
   = note: Undefined symbols for architecture arm64:
             "dyld_stub_binder", referenced from:
                 <initial-undefines>
           ld: symbol(s) not found for architecture arm64
           clang: error: linker command failed with exit code 1 (use -v to see invocation)

error: aborting due to 1 previous error; 29 warnings emitted
------------------------------------------



---- [ui] tests/ui/traits/const-traits/effects/minicore-deref-fail.rs stdout ----

error: auxiliary build of "/Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs" failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" "/Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/work/rust/rust/vendor" "--sysroot" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2" "--target=aarch64-apple-darwin" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-Znext-solver" "--crate-type" "dylib" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary"
--- stderr -------------------------------
warning: unused variable: `rhs`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:39:18
   |
---

warning: unused variable: `x`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:45:9
   |
LL |     let x = 42_i32 + 43_i32;
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: unused variable: `x`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:49:9
   |
   |
LL |     let x = 42_i32 + 43_i32;
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: unused variable: `bytes`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:140:28
   |
   |
LL |     const fn from_ne_bytes(bytes: [u8; size_of::<Self>()]) -> Self {
   |                            ^^^^^ help: if this is intentional, prefix it with an underscore: `_bytes`
warning: unused variable: `other`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:316:18
   |
   |
LL |     fn eq(&self, other: &str) -> bool {
   |                  ^^^^^ help: if this is intentional, prefix it with an underscore: `_other`
warning: unused variable: `value`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:415:18
   |
   |
LL |     const fn new(value: T) -> RefCell<T> {
   |                  ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
warning: unused variable: `arg`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:466:5
   |
LL |     arg: ARG,
---

warning: unused variable: `called_at_rt`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:468:5
   |
LL |     called_at_rt: G,
   |     ^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_called_at_rt`

warning: type `ControlFlow<<Self as Try>::Residual, <Self as Try>::Output>` is more private than the item `Try::branch`
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:62:5
   |
LL |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method `Try::branch` is reachable at visibility `pub`
   |
note: but type `ControlFlow<<Self as Try>::Residual, <Self as Try>::Output>` is only usable at visibility `pub(crate)`
  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:71:1
LL | enum ControlFlow<B, C = ()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `#[warn(private_interfaces)]` on by default

---
   |
LL | const fn bar() {
   |          ^^^

warning: variants `Continue` and `Break` are never constructed
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:73:5
LL | enum ControlFlow<B, C = ()> {
   |      ----------- variants in this enum
   |      ----------- variants in this enum
LL |     #[lang = "Continue"]
LL |     Continue(C),
   |     ^^^^^^^^
LL |     #[lang = "Break"]
LL |     Break(B),

warning: function `size_of` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:130:10
   |
   |
LL | const fn size_of<T>() -> usize {


warning: associated functions `repeat_u8` and `from_ne_bytes` are never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:136:14
LL | impl usize {
   | ---------- associated functions in this implementation
LL |     #[rustc_allow_incoherent_impl]
LL |     #[rustc_allow_incoherent_impl]
LL |     const fn repeat_u8(x: u8) -> usize {
...
...
LL |     const fn from_ne_bytes(bytes: [u8; size_of::<Self>()]) -> Self {

warning: function `panic_display` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:146:10
   |
---
   |      ------ variants in this enum
LL |     #[lang = "None"]
LL |     None,
   |     ^^^^
LL |     #[lang = "Some"]
LL |     Some(T),

warning: methods `as_ref` and `as_mut` are never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:231:14
   |
   |
LL | impl<T> Option<T> {
   | ----------------- methods in this implementation
LL |     const fn as_ref(&self) -> Option<&T> {
...
LL |     const fn as_mut(&mut self) -> Option<&mut T> {
   |              ^^^^^^


warning: function `as_deref` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:248:10
   |
LL | const fn as_deref<T>(opt: &Option<T>) -> Option<&T::Target>

warning: enum `Result` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:283:6
   |
---
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:371:14
   |
LL | impl<T> Option<T> {
   | ----------------- methods in this implementation
LL |     const fn as_pin_ref(self: Pin<&Self>) -> Option<Pin<&T>> {
...
...
LL |     const fn as_pin_mut(self: Pin<&mut Self>) -> Option<Pin<&mut T>> {

warning: struct `RefCell` is never constructed
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:410:8
   |
---
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:415:14
   |
LL | impl<T> RefCell<T> {
   | ------------------ associated function in this implementation
LL |     const fn new(value: T) -> RefCell<T> {

warning: method `borrow` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:420:8
   |
---
LL | struct Ref<'b, T: ?Sized + 'b> {
   |        --- fields in this struct
LL |     value: *const T,
   |     ^^^^^
LL |     borrow: &'b UnsafeCell<()>,

warning: function `const_eval_select` is never used
##[warning]  --> /Users/runner/work/rust/rust/tests/ui/traits/const-traits/effects/auxiliary/minicore.rs:465:10
   |
   |
LL | const fn const_eval_select<ARG: Tuple, F, G, RET>(

error: linking with `cc` failed: exit status: 1
   |
   |
   = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/bin:/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin:/Users/runner/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/runner/.cargo/bin:/usr/local/opt/curl/bin:/usr/local/bin:/usr/local/sbin:/Users/runner/bin:/Users/runner/.yarn/bin:/Users/runner/Library/Android/sdk/tools:/Users/runner/Library/Android/sdk/platform-tools:/Library/Frameworks/Python.framework/Versions/Current/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/bin:/bin:/usr/sbin:/sbin:/Users/runner/.dotnet/tools" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-Wl,-exported_symbols_list" "-Wl,/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustcgWQRtS/list" "/var/folders/g6/rgtlsw6n123b0gt5483s5_cm0000gn/T/rustcgWQRtS/symbols.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary/minicore.minicore.356631b40ca0b657-cgu.0.rcgu.o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary/minicore.0lca3mr32tz9pvmmfm5apoe3h.rcgu.rmeta" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-L" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary" "-o" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/ui/traits/const-traits/effects/minicore-deref-fail/auxiliary/libminicore.dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-install_name" "-Wl,@rpath/libminicore.dylib" "-nodefaultlibs"
   = note: Undefined symbols for architecture arm64:
             "dyld_stub_binder", referenced from:
                 <initial-undefines>
           ld: symbol(s) not found for architecture arm64
           clang: error: linker command failed with exit code 1 (use -v to see invocation)

error: aborting due to 1 previous error; 29 warnings emitted
------------------------------------------

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11649791276/job/32438212018)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11649791276/job/32438212018","builder_name":"checks-actions"} -->

compiler-errors commented:
@bors try

bors commented:
:hourglass: Trying commit b9bab185764b67350b11114d03ecca2e7305a83d with merge 62916d6637a8068fbbe09d7733d71c845b998511...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b9bab185764b67350b11114d03ecca2e7305a83d","merge_sha":"62916d6637a8068fbbe09d7733d71c845b998511"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11653863682/job/32447625683)
Build commit: 62916d6637a8068fbbe09d7733d71c845b998511 (`62916d6637a8068fbbe09d7733d71c845b998511`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11653863682/job/32447625683"},"merge_sha":"62916d6637a8068fbbe09d7733d71c845b998511"} -->

compiler-errors commented:
cool, looks like i was just missing a `no-prefer-dynamic` on the minicore.

compiler-errors commented:
Rebased due to #132479.

@bors r=fee1-dead

bors commented:
:pushpin: Commit d860c8be39aaba35e1b48034af4a3f9ac765b761 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead d860c8be39aaba35e1b48034af4a3f9ac765b761 -->
<!-- homu: {"type":"Approved","sha":"d860c8be39aaba35e1b48034af4a3f9ac765b761","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors rollup=maybe

compiler-errors commented:
:( looks like there's still a conflict

compiler-errors commented:
@bors r-

bors commented:
:umbrella: The latest upstream changes (presumably #132581) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
re-rebased

@bors r=fee1-dead

bors commented:
:pushpin: Commit e3bd6b27f3e788c20598cae3aa57a914142606e4 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead e3bd6b27f3e788c20598cae3aa57a914142606e4 -->
<!-- homu: {"type":"Approved","sha":"e3bd6b27f3e788c20598cae3aa57a914142606e4","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e3bd6b27f3e788c20598cae3aa57a914142606e4 with merge 432972cae64d736b892e7a4c8b4fe7fe0e888904...
<!-- homu: {"type":"BuildStarted","head_sha":"e3bd6b27f3e788c20598cae3aa57a914142606e4","merge_sha":"432972cae64d736b892e7a4c8b4fe7fe0e888904"} -->

rust-timer commented:
Finished benchmarking commit (432972cae64d736b892e7a4c8b4fe7fe0e888904): [comparison URL](https://perf.rust-lang.org/compare.html?start=ca87b535a05097df6abbe2a031b057de2cefac5b&end=432972cae64d736b892e7a4c8b4fe7fe0e888904&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ca87b535a05097df6abbe2a031b057de2cefac5b&end=432972cae64d736b892e7a4c8b4fe7fe0e888904&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.3%, 0.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ca87b535a05097df6abbe2a031b057de2cefac5b&end=432972cae64d736b892e7a4c8b4fe7fe0e888904&stat=max-rss)
<details>
<summary>Results (secondary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ca87b535a05097df6abbe2a031b057de2cefac5b&end=432972cae64d736b892e7a4c8b4fe7fe0e888904&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=ca87b535a05097df6abbe2a031b057de2cefac5b&end=432972cae64d736b892e7a4c8b4fe7fe0e888904&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 779.908s -> 777.534s (-0.30%)
**Artifact size**: 335.29 MiB -> 335.28 MiB (-0.00%)
